### PR TITLE
Make engine version optional

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Fixes this issue:
- #7917

This is caused by the dashboard expecting `engineVersion` to be present, whereas it is actually optional.

### Important Notes
This may have been fixed on the Project Manager side, so the bug may not be reproducible. To properly test that this fix works, an older `project-manager` (e.g `2023.2.1-nightly.2023.9.30`) should be used.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts [hunk 1]
```diff
@@ -414,4 +412,24 @@ export class LocalBackend extends backend.Backend {
...
+    /** Get a fallback engine version. */
+    protected async getFallbackEngineVersion(): Promise<string> {
+        if (this.fallbackEngineVersion != null) {
+            return this.fallbackEngineVersion
+        } else {
+            const versions = await this.listVersions({
+                versionType: backend.VersionType.backend,
+                default: true,
+            })
+            const version = versions[versions.length - 1]
```
Comment (JaroslavTulach) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: What if the `versions` array is zero length?

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts [hunk 2]
```diff
@@ -414,4 +412,24 @@ export class LocalBackend extends backend.Backend {
     override deleteTag() {
         return Promise.resolve()
     }
+
+    // === Protected methods ===
+
+    /** Get a fallback engine version. */
+    protected async getFallbackEngineVersion(): Promise<string> {
+        if (this.fallbackEngineVersion != null) {
```
Comment (JaroslavTulach) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: Shalln't `fallbackEngineVersion` be `Promise<string>` itself and just `resolve` itself to either `listVersions[0]` or some default value? This kind of _promise that reads a this field_ makes it hard to reason about the real flow of the code, in my opinion.

**Review by JaroslavTulach — APPROVED** [review]
No tests? Let's hope this helps. I'll check when playing with next bookclub.

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts (hunk 1 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: then it'll be out of bounds so it'll be `undefined` - the condition below checks for that case, and throws if `version` is `undefined` (meaning `versions` was zero length)

Comment (JaroslavTulach) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: Ááá, JavaScript:
```
> undefined == null
true
```

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: good point, should now be addressed? although all i did was inline the method into the property, not sure how much that helps

**Comment by somebody1234** [comment]
did some further testing and fixed some bugs

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts (hunk 2 above)
Comment (JaroslavTulach) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: 1. I can read it
2. the state management is simplified
3. the `Future` is computed once and then returns always the same value
4. that wasn't clear to me from the previous version

**Review by 4e6 — APPROVED** [review]

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/projectManager.ts [hunk 3]
```diff
@@ -100,7 +100,7 @@ export interface CreateProject {
 
 /** The return value of the "open project" endpoint. */
 export interface OpenProject {
-    engineVersion: string
+    engineVersion?: string
```
Comment (4e6) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/projectManager.ts]: Engine version is always returned in the [`project/open`](https://github.com/enso-org/enso/blob/f1356374e19878316d53a92b7880d3725d2f32f4/docs/language-server/protocol-project-manager.md#projectopen) response. It was optional, and then removed in the `project/list` response

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts [hunk 4]
```diff
@@ -146,15 +168,13 @@ export class LocalBackend extends backend.Backend {
         if (cachedProject == null) {
             const result = await this.projectManager.listProjects({})
             const project = result.projects.find(listedProject => listedProject.id === projectId)
-            const engineVersion = project?.engineVersion
+            const engineVersion = project?.engineVersion ?? (await this.fallbackEngineVersion)
```
Comment (4e6) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: Why do we need to know the version of the project? Where it is used?

**Review by 4e6 — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts (hunk 4 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/localBackend.ts]: good point... it's only needed for the cloud backend as the location for its wasm bundle depends on the version of the IDE. should now be fixed.

**Review by 4e6 — APPROVED** [review]

**Comment by somebody1234** [comment]
@PabloBuchu @indiv0 ping for visibility

**Comment by PabloBuchu** [comment]
looks like the PR fixed original issue. 👍

**Review by PabloBuchu — APPROVED** [review]

**Pull request merged**